tasmota/tasmota_xsns_sensor/xsns_114_amsx915.ino

```diff
@@ -0,0 +1,217 @@
+        WSContentSend_Temp("AMSx915", amsx915data.temperature);
+#endif  // USE_WEBSERVER
+      }
+  }
+}
+
+/*********************************************************************************************\
+ * Driver Settings load and save
+\*********************************************************************************************/
+
+void Amsx915SettingsLoad(bool erase) {
+  // Called from FUNC_PRE_INIT (erase = 0) once at restart
+  memset(&amsx915Settings, 0x00, sizeof(amsx915Settings));
+  amsx915Settings.version = AMSX915_VERSION;
+  amsx915Settings.pmax = PMAX_DEFAULT;
+  amsx915Settings.pmin = PMIN_DEFAULT;
+
+  // *** End Init default values ***
+
+#ifndef USE_UFILESYS
+  AddLog(LOG_LEVEL_INFO, PSTR("CFG: AMS defaults as file system not enabled"));
+#else
+  // Try to load sensor config file
+  char filename[20];
+  snprintf_P(filename, sizeof(filename), PSTR(TASM_FILE_SENSOR), XSNS_114);
+
+  if (erase) {
+    TfsDeleteFile(filename);  // Use defaults
+  }
+  else if (TfsLoadFile(filename, (uint8_t*)&amsx915Settings, sizeof(amsx915Settings))) {
+    if (amsx915Settings.version != AMSX915_VERSION) {      // Fix version dependent changes
+      // Set current version and save settings
+      amsx915Settings.version = AMSX915_VERSION;
+      Amsx915SettingsSave();
+    }
+    AddLog(LOG_LEVEL_DEBUG, PSTR("CFG: AMS config loaded from file"));
+  }
+  else {
+    // File system not ready: No flash space reserved for file system
+    AddLog(LOG_LEVEL_DEBUG, PSTR("CFG: AMS use defaults as file system not ready or file not found"));
+  }
+#endif  // USE_UFILESYS
+}
+
+void Amsx915SettingsSave(void) {
+  // Called from FUNC_SAVE_SETTINGS every SaveData second and at restart
+#ifdef USE_UFILESYS
+  uint32_t crc32 = GetCfgCrc32((uint8_t*)&amsx915Settings +4, sizeof(amsx915Settings) -4);  // Skip crc32
+  if (crc32 != amsx915Settings.crc32) {
+    // Try to save sensor config file
+    amsx915Settings.crc32 = crc32;
+
+    char filename[20];
+    snprintf_P(filename, sizeof(filename), PSTR(TASM_FILE_SENSOR), XSNS_114);
+
+    if (TfsSaveFile(filename, (const uint8_t*)&amsx915Settings, sizeof(amsx915Settings))) {
+      AddLog(LOG_LEVEL_DEBUG, PSTR("CFG: AMS Settings saved to file"));
+    } else {
+      // File system not ready: No flash space reserved for file system
+      AddLog(LOG_LEVEL_DEBUG, PSTR("CFG: ERROR AMS file system not ready or unable to save file"));
+    }
+  }
+#endif  // USE_UFILESYS
+}
+
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+
+bool Xsns114(uint32_t function) {
+  if (!I2cEnabled(XI2C_86)) { return false; }
+
+  bool result = false;
+
+  switch(function) {
+    case FUNC_PRE_INIT:
+      Amsx915SettingsLoad(0);
+      break;
+    case FUNC_INIT:
+      Amsx915ReadData();
```

Comment:
I'm still not ok with this. You never check if the device is actually plugged into the bus, and blindly send requests to the bus every second. We can't accept this kind of behavior from every driver, it would overwelm the bus and take a long time for each I2C request to time out when the device is not present.

Reply:
See other drivers such as the one I pointed above for rules of good design